Replace List Mesh to Tensor

jjiangTT opened the pull request:
### Ticket
Link to Github Issue
https://github.com/tenstorrent/tt-metal/issues/15061

### Problem description
ListMeshToTensor was a python class in distributed.py that didn't match the xtensor->torch.tensor convention and instead output a list[torch.tensor]. I've added the utility method ttnn.shardedtensor_to_tensorlist(ttnn.tensor)->list[torch.tensor] instead.

### What's changed
-ListMeshToTensor removed, all usages replaced with ttnn.sharded_tensor_to_torch_tensor_list (tensor: ttnn.tensor) hook
-Added temporary python hook in operations/core.py to convert from ttnn.tensor lists to torch.tensor lists

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/13335705560
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [x] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes: [tests/ttnn/unit_tests/test_multi_device_async.py](https://github.com/tenstorrent/tt-metal/blob/8063f0ab02bec18561f9461e8d1ff868ba0eeaf0/tests/ttnn/unit_tests/test_multi_device_async.py#L87), [tests/ttnn/unit_tests/operations/test_creation.py](https://github.com/tenstorrent/tt-metal/blob/8063f0ab02bec18561f9461e8d1ff868ba0eeaf0/tests/ttnn/unit_tests/operations/test_creation.py#L261), [tests/ttnn/unit_tests/tensor/test_tensor_prealloc_and_write.py](https://github.com/tenstorrent/tt-metal/blob/8063f0ab02bec18561f9461e8d1ff868ba0eeaf0/tests/ttnn/unit_tests/tensor/test_tensor_prealloc_and_write.py#L82), [tests/ttnn/distributed/test_multidevice_TG.py](https://github.com/tenstorrent/tt-metal/blob/8063f0ab02bec18561f9461e8d1ff868ba0eeaf0/tests/ttnn/distributed/test_multidevice_TG.py#L19), [models/demos/t3000/falcon40b/tests/test_falcon_model.py](https://github.com/tenstorrent/tt-metal/blob/8063f0ab02bec18561f9461e8d1ff868ba0eeaf0/models/demos/t3000/falcon40b/tests/test_falcon_model.py#L9)

jjiangTT committed:
initial implementation for new helper method and replacing all paths across codebase

jjiangTT committed:
change assert to tt_throw to satisfy compiler

jjiangTT committed:
change assert to tt_throw to satisfy compiler

jjiangTT committed:
add get_mesh_device_core_grid

jjiangTT committed:
add get_mesh_device_core_grid

jjiangTT committed:
fix module names and variant unpacking

jjiangTT committed:
fix module names and variant unpacking

jjiangTT committed:
added rdoc, fixed bfp_b type error by adding conversion

jjiangTT committed:
rm -rf ListMeshToTensor

jjiangTT committed:
Bash remaining imports and usages

jjiangTT committed:
remove legacyshape, typecheck only once for original tensor and pass types forward, fix bfloat16 type error

jjiangTT committed:
attempted fix for devicestorage error

jjiangTT committed:
fixed devicestorage, bfp_b conversion errors

jjiangTT committed:
add working python hook for sharded_tensor_to_tensor_list,add alias for ttnn.multidevice, fix python hook name recursion error

omilyutin-tt commented on ttnn/cpp/ttnn/distributed/distributed_pybind.cpp:
```diff
@@ -308,37 +340,73 @@ void py_module(py::module& module) {
         [](const std::vector<Tensor>& tensors) -> Tensor { return aggregate_as_tensor(tensors, AllGatherTensor{}); },
         py::arg("tensors"),
         py::kw_only());
+    module.def(
+        "get_mesh_device_core_grid",
+        [](MeshDevice& mesh_device) {
+            CoreCoord coords = mesh_device.compute_with_storage_grid_size();
+            new CoreGrid(coords.x, coords.y);
+        },
```
Is this new / needed? Doesn't `new CoreGrid` result in a memory leak?

jjiangTT replied:
Not needed for this pr, I'll remove it - was one of the functions in distributed.py.

omilyutin-tt replied:
Ok. Btw I think it is fine to leave some stuff in distributed.py - in particular the visualization seems like only makes sense for REPL use case. Other functions just wrap C++ methods anyways.

omilyutin-tt commented on ttnn/ttnn/__init__.py:
```diff
@@ -281,6 +286,15 @@ def auto_register_ttnn_cpp_operations(module):
+_original_sharded_tensor_to_tensor_list = ttnn.sharded_tensor_to_tensor_list
+
+
+def sharded_tensor_to_tensor_list_helper(tensor: ttnn.Tensor) -> List["torch.Tensor"]:
+    return [ttnn.to_torch(shard) for shard in ttnn._original_sharded_tensor_to_tensor_list(tensor)]
```
I think we will end up defining a Python function regardless, so would something like this work:
```
def sharded_tensor_to_tensor_list_helper(tensor: ttnn.Tensor) -> List["torch.Tensor"]:
  tensor_shards = get_device_tensors(tensor.cpu())
  return [ttnn.to_torch(shard) for shard in tensor_shards]
  
```

omilyutin-tt replied:
I'd also keep `to_torch` in the name somewhere, something like `to_torch_sharded`? @cfjchu wdyt? Need to add a function documentation that explains we are converting multi-device TTNN tensor shards into a list of torch tensors.

jjiangTT replied:
> I think we will end up defining a Python function regardless, so would something like this work:
> 
> ```
> def sharded_tensor_to_tensor_list_helper(tensor: ttnn.Tensor) -> List["torch.Tensor"]:
>   tensor_shards = get_device_tensors(tensor.cpu())
>   return [ttnn.to_torch(shard) for shard in tensor_shards]
>   
> ```

I'm planning on going through and reworking my original code to fit in over the weekend and then moving the function completely to c++ since I think the performance would be slightly better and it also seems a little cleaner to me from an api standpoint. This is also partly since I think the to_torch method will likely be used for some of my other PRs relating to this task (the other mesh conversion classes). 
Should there be no concerns my intention was to just add this hook as an interim measure in the meantime so that all I have to do later is change things on that end, delete this hook, then rebuild for it to show up in python - everything there was already running and passing test cases so all I would have to do is refactor the codebase a bit.

omilyutin-tt replied:
I don't think you are winning much in terms of performance, as you still need to jump into Python to call `to_torch`? All the heavy lifting will be done by C++ still, so we should be good. In general, let's avoid optimizing for performance unless there is a reason to.

Re API stand point. The APIs in both cases are identical though, right? Just a function `sharded_tensor_to_tensor_list_helper` from the user's perspective. I'm more concerned about adding new code that is unnecessary.

> This is also partly since I think the to_torch method will likely be used for some of my other PRs relating to this task (the other mesh conversion classes).

Perhaps let's sync on the exact plan. My understanding was that you don't need to add anything in C++, except for exposing the mesh mappers / composers into Python. We will run the model perf tests and see if we need to optimize the code further, but for now I'd opt for the simple approach.

jjiangTT replied:
Update: we discussed over sync and a lot of things got clarified for me.

jjiangTT committed:
Remove unfinished method get_mesh_device_core_grid

    module.def(
        "get_mesh_device_core_grid",
        [](MeshDevice& mesh_device) {
            CoreCoord coords = mesh_device.compute_with_storage_grid_size();
            new CoreGrid(coords.x, coords.y);
        },
        py::arg("mesh_device"));

jjiangTT committed:
simplify and rename sharded_tensor_to_torch_tensor_list, move it to ttnn/operations/core.py

jjiangTT committed:
update pydoc with example, args

jjiangTT committed:
fix name conflict

jjiangTT committed:
fix name conflict

jjiangTT committed:
naming fix

jjiangTT committed:
remove unused import causing errors

jjiangTT committed:
actually rm-list_mesh_to_tensor, change all 24 usages over to a one liner array generator

omilyutin-tt commented:
The PR looks good! Make sure to kick off post commits and T3K unit tests to double check nothing is broken.

cfjchu commented on ttnn/ttnn/__init__.py:
```diff
@@ -9,8 +9,10 @@
 import pathlib
 import re
 from types import ModuleType
+from typing import List
```
Please remove all added lines here.

cfjchu commented on ttnn/cpp/ttnn/distributed/distributed_pybind.hpp:
```diff
@@ -4,6 +4,9 @@
 
 #pragma once
 #include "pybind11/pybind_fwd.hpp"
+#include <pybind11/operators.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
```
You shouldn't need these here. Please remove.

cfjchu commented on ttnn/cpp/ttnn/distributed/distributed_pybind.cpp:
```diff
@@ -3,14 +3,30 @@
 
 #include "ttnn/distributed/distributed_pybind.hpp"
+#include <pybind11/pytypes.h>
+#include <cstdint>
+#include <memory>
 #include <utility>
 
+#include "tt-metalium/assert.hpp"
+#include "tt-metalium/bfloat16.hpp"
+#include "tt-metalium/buffer.hpp"
+#include "tt-metalium/core_coord.hpp"
+#include "tt-metalium/overloaded.hpp"
+#include "tt-metalium/core_coord.hpp"
+#include "ttnn/common/constants.hpp"
 #include "ttnn/distributed/api.hpp"
+#include "ttnn/tensor/host_buffer/borrowed_buffer.hpp"
+#include "ttnn/tensor/host_buffer/functions.hpp"
+#include "ttnn/tensor/layout/page_config.hpp"
+#include "ttnn/tensor/storage.hpp"
+#include "ttnn/tensor/tensor_impl.hpp"
 #include "ttnn/tensor/tensor_utils.hpp"
 #include "ttnn/tensor/tensor.hpp"
 #include "ttnn/types.hpp"
 #include <tt-metalium/command_queue.hpp>
-#include "pybind11/stl.h"
+
+#include "umd/device/tt_xy_pair.h"
```
You've added a lot of new includes. Can you please make sure you're only including the ones you need?

cfjchu commented:
Thanks, @jjiangTT! Just need to clean up some of the includes and then we can merge this in.

jjiangTT committed:
remove unused imports

jjiangTT committed:
add back some includes

jjiangTT committed:
fix no newline at end of file
